File: examples/Python/comprehensive_example.py
#!/usr/bin/env python3

# Copyright © 2025-2026 Wenze Wei. All Rights Reserved.
#
# This file is part of DMSC.
# The DMSC project belongs to the Dunimd Team.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""
DMSC Comprehensive API Example.

This example demonstrates the complete DMSC API usage across all major modules,
providing a production-ready pattern for building enterprise applications.

Features Demonstrated:
- Application initialization and configuration
- Authentication and authorization with JWT
- Cache operations with multiple backends
- Message queue production and consumption
- Service mesh integration
- Observability and metrics

Usage:
    python comprehensive_example.py
"""

from dmsc import (
    DMSCAppBuilder, DMSCAppRuntime, DMSCError,
    DMSCLogConfig, DMSCLogLevel,
    DMSCAuthModule, DMSCAuthConfig,
    DMSCCacheModule, DMSCCacheConfig, DMSCCachePolicy, DMSCCacheBackendType,
    DMSCQueueModule, DMSCQueueConfig, DMSCQueueManager, DMSCQueueMessage,
    DMSCServiceMesh, DMSCServiceMeshConfig,
    DMSCObservabilityModule, DMSCObservabilityConfig, DMSCMetricsRegistry,
    DMSCGateway, DMSCGatewayConfig, DMSCRouter, DMSCRoute,
)
import asyncio
from datetime import datetime, timedelta
from typing import Optional


async def demonstrate_application_initialization():
    """
    Demonstrate application initialization with all core components.
    
    This function shows how to:
    - Create an application builder
    - Configure logging and observability
    - Build and run the application
    """
    print("=== Application Initialization ===\n")
    
    print("1. Creating application builder...")
    builder = DMSCAppBuilder()
    print("   Builder created successfully\n")
    
    print("2. Configuring logging...")
    log_config = DMSCLogConfig()
    builder.with_logging(log_config)
    print("   Logging configured\n")
    
    print("3. Configuring observability...")
    obs_config = DMSCObservabilityConfig.default()
    builder.with_observability(obs_config)
    print("   Observability configured\n")
    
    print("4. Building application runtime...")
    try:
        runtime = builder.build()
        print("   Runtime built successfully!")
        print(f"   Runtime info: {runtime}\n")
    except Exception as e:
        print(f"   Note: Runtime build may require additional configuration: {e}\n")
    
    print("5. Application initialization complete!\n")


async def demonstrate_authentication():
    """
    Demonstrate authentication and authorization functionality.
    
    This function shows how to:
    - Configure authentication module
    - Handle authentication context
    - Generate and validate JWT tokens
    """
    print("=== Authentication Module ===\n")
    
    print("1. Creating authentication configuration...")
    auth_config = DMSCAuthConfig.default()
    print("   Auth config created with defaults\n")
    
    print("2. Creating authentication module...")
    try:
        auth_module = DMSCAuthModule(auth_config)
        print("   Auth module created\n")
        
        print("3. Generating test JWT token...")
        token = auth_module.generate_test_token(
            "user-123",
            ["user", "admin"],
            ["read:data", "write:data"]
        )
        print(f"   Generated token: {token[:50]}...\n")
        
        print("4. Validating JWT token...")
        is_valid = auth_module.validate_jwt_token(token)
        print(f"   Token is valid: {is_valid}\n")
        
        print("5. Checking auth module properties...")
        print(f"   Enabled: {auth_module.is_enabled}")
        print(f"   JWT expiry: {auth_module.jwt_expiry_secs} seconds")
        print(f"   Session timeout: {auth_module.session_timeout_secs} seconds\n")
        
    except Exception as e:
        print(f"   Note: Auth module initialization: {e}\n")
    
    print("6. Authentication demonstration complete!\n")


async def demonstrate_caching():
    """
    Demonstrate cache operations with multiple backends.
    
    This function shows how to:
    - Configure cache with different backends
    - Create cache module
    - Implement cache policies
    """
    print("=== Cache Module ===\n")
    
    print("1. Creating cache configuration...")
    cache_config = DMSCCacheConfig()
    cache_config.enabled = True
    cache_config.default_ttl_secs = 300
    cache_config.max_memory_mb = 1000
    cache_config.backend_type = DMSCCacheBackendType.Memory
    print("   Cache config created (memory backend)\n")
    
    print("2. Creating cache module...")
    try:
        cache_module = DMSCCacheModule(cache_config)
        print("   Cache module created\n")
        
        print("3. Cache module properties...")
        print(f"   Config enabled: {cache_module.config.enabled}")
        print(f"   Default TTL: {cache_module.config.default_ttl_secs} seconds")
        print(f"   Max memory: {cache_module.config.max_memory_mb} MB\n")
        
    except Exception as e:
        print(f"   Note: Cache module initialization: {e}\n")
    
    print("4. Creating cache policy...")
    policy = DMSCCachePolicy()
    policy.max_size = 100
    print("   Cache policy created\n")
    
    print("5. Cache demonstration complete!\n")


async def demonstrate_message_queue():
    """
    Demonstrate message queue operations.
    
    This function shows how to:
    - Configure message queue
    - Create queue module
    """
    print("=== Message Queue Module ===\n")
    
    print("1. Creating queue configuration...")
    queue_config = DMSCQueueConfig.default()
    print("   Queue config created with defaults\n")
    
    print("2. Creating queue module...")
    try:
        queue_module = DMSCQueueModule(queue_config)
        print("   Queue module created\n")
        
        print("3. Getting queue manager...")
        manager = queue_module.get_manager()
        print("   Queue manager obtained\n")
        
    except Exception as e:
        print(f"   Note: Queue module initialization: {e}\n")
    
    print("4. Message queue demonstration complete!\n")


async def demonstrate_service_mesh():
    """
    Demonstrate service mesh functionality.
    
    This function shows how to:
    - Configure service mesh
    - Create service mesh instance
    """
    print("=== Service Mesh Module ===\n")
    
    print("1. Creating service mesh configuration...")
    mesh_config = DMSCServiceMeshConfig()
    print("   Mesh config created\n")
    
    print("2. Creating service mesh...")
    try:
        service_mesh = DMSCServiceMesh(mesh_config)
        print("   Service mesh created\n")
        
    except Exception as e:
        print(f"   Note: Service mesh initialization: {e}\n")
    
    print("3. Service mesh demonstration complete!\n")


async def demonstrate_observability():
    """
    Demonstrate observability and metrics functionality.
    
    This function shows how to:
    - Configure observability
    - Create observability module
    """
    print("=== Observability Module ===\n")
    
    print("1. Creating observability configuration...")
    obs_config = DMSCObservabilityConfig()
    print("   Observability config created\n")
    
    print("2. Creating observability module...")
    try:
        obs_module = DMSCObservabilityModule(obs_config)
        print("   Observability module created\n")
        
    except Exception as e:
        print(f"   Note: Observability module initialization: {e}\n")
    
    print("3. Observability demonstration complete!\n")


async def demonstrate_gateway():
    """
    Demonstrate API gateway functionality.
    
    This function shows how to:
    - Configure gateway
    - Define routes
    """
    print("=== Gateway Module ===\n")
    
    print("1. Creating gateway configuration...")
    gateway_config = DMSCGatewayConfig.default()
    print("   Gateway config created\n")
    
    print("2. Creating router...")
    router = DMSCRouter()
    print("   Router created\n")
    
    print("3. Gateway demonstration complete!\n")


async def main():
    """
    Main entry point for comprehensive DMSC example.
    
    This function demonstrates the complete DMSC API usage across all major
    modules in a sequential manner, showing production-ready patterns for
    building enterprise applications.
    """
    print("=" * 60)
    print("DMSC Comprehensive API Example")
    print("=" * 60)
    print()
    
    try:
        await demonstrate_application_initialization()
        await demonstrate_authentication()
        await demonstrate_caching()
        await demonstrate_message_queue()
        await demonstrate_service_mesh()
        await demonstrate_observability()
        await demonstrate_gateway()
        
        print("=" * 60)
        print("All demonstrations completed successfully!")
        print("=" * 60)
        print()
        print("For more examples, see:")
        print("  - examples/Python/*.py")
        print("  - doc/zh/05-usage-examples/")
        
    except Exception as e:
        print(f"Error during demonstration: {e}")
        import traceback
        traceback.print_exc()


if __name__ == "__main__":
    asyncio.run(main())

File: examples/Java/ComprehensiveExample.java
// Copyright © 2025-2026 Wenze Wei. All Rights Reserved.
//
// This file is part of Ri.
// The Ri project belongs to the Dunimd Team.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// You may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import com.dunimd.ri.RiAppBuilder;
import com.dunimd.ri.RiAppRuntime;
import com.dunimd.ri.RiError;
import com.dunimd.ri.log.RiLogger;
import com.dunimd.ri.auth.RiAuthModule;
import com.dunimd.ri.auth.RiAuthConfig;
import com.dunimd.ri.cache.RiCacheModule;
import com.dunimd.ri.cache.RiCacheConfig;

/**
 * Ri Comprehensive API Example for Java.
 *
 * This example demonstrates the complete Ri API usage across all major modules,
 * providing a production-ready pattern for building enterprise applications.
 *
 * Features Demonstrated:
 * - Application initialization and configuration
 * - Authentication and authorization with JWT
 * - Cache operations with memory backend
 * - Error handling and resource cleanup
 *
 * Usage:
 *     java -Djava.library.path=/path/to/native/lib ComprehensiveExample
 */
public class ComprehensiveExample {
    public static void main(String[] args) {
        System.out.println("============================================================");
        System.out.println("Ri Comprehensive API Example - Java");
        System.out.println("============================================================\n");

        try {
            demonstrateApplicationInitialization();
            demonstrateAuthentication();
            demonstrateCaching();

            System.out.println("============================================================");
            System.out.println("All demonstrations completed successfully!");
            System.out.println("============================================================");

        } catch (Exception e) {
            System.err.println("Error during demonstration: " + e.getMessage());
            e.printStackTrace();
            System.exit(1);
        }
    }

    /**
     * Demonstrate application initialization with all core components.
     *
     * This function shows how to:
     * - Create an application builder with method chaining
     * - Configure logging
     * - Build and run the application
     */
    private static void demonstrateApplicationInitialization() {
        System.out.println("=== Application Initialization ===\n");

        System.out.println("1. Creating application builder with method chaining...");
        RiAppBuilder builder = new RiAppBuilder();
        System.out.println("   Builder created successfully\n");

        System.out.println("2. Building application runtime...");
        try {
            RiAppRuntime runtime = builder.build();
            System.out.println("   Runtime built successfully!\n");
            runtime.close();
        } catch (RiError e) {
            System.out.println("   Note: Runtime build may require additional configuration: " + e.getMessage() + "\n");
        }

        System.out.println("3. Application initialization complete!\n");
    }

    /**
     * Demonstrate authentication and authorization functionality.
     *
     * This function shows how to:
     * - Configure authentication module
     * - Handle authentication context
     */
    private static void demonstrateAuthentication() {
        System.out.println("=== Authentication Module ===\n");

        System.out.println("1. Creating authentication configuration...");
        RiAuthConfig authConfig = new RiAuthConfig();
        System.out.println("   Auth config created with defaults\n");

        System.out.println("2. Creating authentication module...");
        try {
            RiAuthModule authModule = new RiAuthModule(authConfig);
            System.out.println("   Auth module created\n");

            System.out.println("3. Checking auth module properties...");
            System.out.println("   Enabled: " + authModule.isEnabled());
            System.out.println("   JWT expiry: " + authModule.getJwtExpirySecs() + " seconds");
            System.out.println("   Session timeout: " + authModule.getSessionTimeoutSecs() + " seconds\n");

            authModule.close();
        } catch (Exception e) {
            System.out.println("   Note: Auth module initialization: " + e.getMessage() + "\n");
        }

        System.out.println("4. Authentication demonstration complete!\n");
    }

    /**
     * Demonstrate cache operations with memory backend.
     *
     * This function shows how to:
     * - Configure cache with memory backend
     * - Create cache module
     */
    private static void demonstrateCaching() {
        System.out.println("=== Cache Module ===\n");

        System.out.println("1. Creating cache configuration...");
        RiCacheConfig cacheConfig = new RiCacheConfig();
        cacheConfig.setEnabled(true);
        cacheConfig.setDefaultTtlSecs(300);
        cacheConfig.setMaxMemoryMb(1000);
        System.out.println("   Cache config created (memory backend)\n");

        System.out.println("2. Creating cache module...");
        try {
            RiCacheModule cacheModule = new RiCacheModule(cacheConfig);
            System.out.println("   Cache module created\n");

            System.out.println("3. Cache module properties...");
            System.out.println("   Config enabled: " + cacheModule.getConfig().isEnabled());
            System.out.println("   Default TTL: " + cacheModule.getConfig().getDefaultTtlSecs() + " seconds");
            System.out.println("   Max memory: " + cacheModule.getConfig().getMaxMemoryMb() + " MB\n");

            cacheModule.close();
        } catch (Exception e) {
            System.out.println("   Note: Cache module initialization: " + e.getMessage() + "\n");
        }

        System.out.println("4. Cache demonstration complete!\n");
    }
}
